### PR TITLE
Tweak suggestions for bare trait used as a type

### LINKED ISSUE #119525: rustc suggests `dyn SomeTrait` for traits which aren't object safe
### Code

```Rust
fn ord_prefer_dot(s: &str) -> Ord {
    (s.starts_with("."), s)
}
```

### Current output

```Shell
error[E0038]: the trait `Ord` cannot be made into an object
   --> src/lib.rs:1:31
    |
1   | fn ord_prefer_dot(s: &str) -> Ord {
    |                               ^^^ `Ord` cannot be made into an object
    |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
   --> /playground/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/cmp.rs:305:15
    |
305 | pub trait Eq: PartialEq<Self> {
    |               ^^^^^^^^^^^^^^^ the trait cannot be made into an object because it uses `Self` as a type parameter
...
793 | pub trait Ord: Eq + PartialOrd<Self> {
    |                     ^^^^^^^^^^^^^^^^ the trait cannot be made into an object because it uses `Self` as a type parameter

error[E0782]: trait objects must include the `dyn` keyword
 --> src/lib.rs:1:31
  |
1 | fn ord_prefer_dot(s: &str) -> Ord {
  |                               ^^^
  |
help: add `dyn` keyword before this trait
  |
1 | fn ord_prefer_dot(s: &str) -> dyn Ord {
  |                               +++

Some errors have detailed explanations: E0038, E0782.
For more information about an error, try `rustc --explain E0038`.
```

### Desired output

```Shell
error[E0038]: the trait `Ord` cannot be made into an object
   --> src/lib.rs:1:31
    |
1   | fn ord_prefer_dot(s: &str) -> Ord {
    |                               ^^^ `Ord` cannot be made into an object
    |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
   --> /playground/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/cmp.rs:305:15
    |
305 | pub trait Eq: PartialEq<Self> {
    |               ^^^^^^^^^^^^^^^ the trait cannot be made into an object because it uses `Self` as a type parameter
...
793 | pub trait Ord: Eq + PartialOrd<Self> {
    |                     ^^^^^^^^^^^^^^^^ the trait cannot be made into an object because it uses `Self` as a type parameter

help: add `impl` keyword before this trait
  |
1 | fn ord_prefer_dot(s: &str) -> impl Ord {
  |                               ++++
```

### Rationale and extra context

In this case, a possible valid implementation would be:

```
fn ord_prefer_dot(s: &str) -> impl Ord + '_ {
    (s.starts_with("."), s)
}
```

It would be an improvement to not suggest `dyn Ord`, but ideally rustc would suggest `impl Ord` I think.

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by estebank**
```
error[E0782]: trait objects must include the `dyn` keyword
  --> $DIR/not-on-bare-trait-2021.rs:11:11
   |
LL | fn bar(x: Foo) -> Foo {
   |           ^^^
   |
help: use a generic type parameter, constrained by the trait `Foo`
   |
LL | fn bar<T: Foo>(x: T) -> Foo {
   |       ++++++++    ~
help: you can also use `impl Foo`, but users won't be able to specify the type paramer when calling the `fn`, having to rely exclusively on type inference
   |
LL | fn bar(x: impl Foo) -> Foo {
   |           ++++
help: alternatively, use a trait object to accept any type that implements `Foo`, accessing its methods at runtime using dynamic dispatch
   |
LL | fn bar(x: &dyn Foo) -> Foo {
   |           ++++

error[E0782]: trait objects must include the `dyn` keyword
  --> $DIR/not-on-bare-trait-2021.rs:11:19
   |
LL | fn bar(x: Foo) -> Foo {
   |                   ^^^
   |
help: use `impl Foo` to return an opaque type, as long as you return a single underlying type
   |
LL | fn bar(x: Foo) -> impl Foo {
   |                   ++++
help: alternatively, you can return an owned trait object
   |
LL | fn bar(x: Foo) -> Box<dyn Foo> {
   |                   +++++++    +
```

Fix #119525:

```

error[E0038]: the trait `Ord` cannot be made into an object
  --> $DIR/bare-trait-dont-suggest-dyn.rs:3:33
   |
LL | fn ord_prefer_dot(s: String) -> Ord {
   |                                 ^^^ `Ord` cannot be made into an object
   |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
  --> $SRC_DIR/core/src/cmp.rs:LL:COL
   |
   = note: the trait cannot be made into an object because it uses `Self` as a type parameter
  ::: $SRC_DIR/core/src/cmp.rs:LL:COL
   |
   = note: the trait cannot be made into an object because it uses `Self` as a type parameter
help: consider using an opaque type instead
   |
LL | fn ord_prefer_dot(s: String) -> impl Ord {
   |                                 ++++
```

### CODE REVIEW

File: tests/ui/traits/bound/not-on-bare-trait.stderr [hunk 1]
```diff
@@ -26,9 +34,26 @@ LL | fn foo(_x: impl Foo + Send) {
...
+   |
+LL | fn bar(_x: (dyn Foo + Send)) {
+   |        ^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `(dyn Foo + Send + 'static)`
+   = help: unsized fn params are gated as an unstable feature
+help: you can use `impl Trait` as the argument type
+   |
+LL | fn bar(_x: impl (dyn Foo + Send)) {
+   |            ++++
```
Comment (estebank) [inline tests/ui/traits/bound/not-on-bare-trait.stderr:51]: rats! Given that we're doing some sketchy string handling already, I don't think this is the only problem we'll face, but I'd rather not add *more* string handling logic to this.

Comment (compiler-errors) [reply tests/ui/traits/bound/not-on-bare-trait.stderr:51]: You have access to the `self_ty: hir::Ty<'tcx>` -- why not just check the `TraitObjectSyntax` here? we shouldn't need any string manipulation here at all.

Comment (estebank) [reply tests/ui/traits/bound/not-on-bare-trait.stderr:51]: This one is from the second commit, not the first one. That one only has the additional `Span` from the `ObligationCauseCode::SizedArgumentType`. We could change that one to instead hold the `HirId` instead... 🤔

Comment (estebank) [reply tests/ui/traits/bound/not-on-bare-trait.stderr]: Done.

File: compiler/rustc_hir_analysis/src/astconv/lint.rs [hunk 2]
```diff
@@ -32,32 +33,120 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+    ) -> Option<Vec<(Span, String)>> {
+        // check if the trait has generics, to make a correct suggestion
+        let param_name = generics.params.next_type_param_name(None);
+
+        let add_generic_sugg = if let Some(span) = generics.span_for_param_suggestion() {
+            (span, format!(", {param_name}: {impl_trait_name}"))
+        } else {
+            (generics.span, format!("<{param_name}: {impl_trait_name}>"))
+        };
+        Some(vec![(self_ty_span, param_name), add_generic_sugg])
```
Comment (davidtwco) [inline compiler/rustc_hir_analysis/src/astconv/lint.rs:73]: Can this just return `Vec<(Span, String)>`?

File: compiler/rustc_hir_analysis/src/astconv/lint.rs [hunk 3]
```diff
@@ -32,32 +33,120 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
                      all types that also implement `{impl_trait_name}`"
                 ),
-                vec![(self_ty.span, param_name), add_generic_sugg],
+                sugg,
                 Applicability::MaybeIncorrect,
             );
         }
     }
 
+    fn generics_suggestion(
```
Comment (davidtwco) [inline compiler/rustc_hir_analysis/src/astconv/lint.rs:59]: nit: more descriptive name

File: compiler/rustc_hir_analysis/src/astconv/lint.rs (hunk 2 above)
Comment (davidtwco) [inline compiler/rustc_hir_analysis/src/astconv/lint.rs:73]: What is the first element of this doing? I might be missing something obvious..?

Comment (estebank) [reply compiler/rustc_hir_analysis/src/astconv/lint.rs:73]:
```
+help: use a new generic type parameter, constrained by `Foo + Send`
+   |
+LL | fn foo<T: Foo + Send>(_x: T) {
+   |       +++++++++++++++     ~ `self_ty_span`, this is the "first" part of the suggestion
            |
            `generics.span`
```

**Review by davidtwco — APPROVED** [review]

**Pull request merged**